### PR TITLE
[SPARK-10827] replace volatile with Atomic* in AppClient.scala.

### PR DESCRIPTION
**Pull request by rxin**
This is a followup for #9317 to replace volatile fields with AtomicBoolean and AtomicReference.

### CODE REVIEW

File: core/src/main/scala/org/apache/spark/deploy/client/AppClient.scala [hunk 1]
```diff
@@ -263,18 +265,18 @@ private[spark] class AppClient(
...
-      if (!alreadyDead) {
+      if (!alreadyDead.get) {
         listener.dead(reason)
-        alreadyDead = true
+        alreadyDead.set(true)
       }
     }
 
     override def onStop(): Unit = {
       if (registrationRetryTimer != null) {
```
Comment (zsxwing) [inline core/src/main/scala/org/apache/spark/deploy/client/AppClient.scala]: registrationRetryTimer.get != null

File: core/src/main/scala/org/apache/spark/deploy/client/AppClient.scala [hunk 2]
```diff
@@ -320,7 +322,7 @@ private[spark] class AppClient(
    */
   def killExecutors(executorIds: Seq[String]): Boolean = {
     if (endpoint != null && appId != null) {
```
Comment (zsxwing) [inline core/src/main/scala/org/apache/spark/deploy/client/AppClient.scala]: endpoint.get != null && appId.get != null

File: core/src/main/scala/org/apache/spark/deploy/client/AppClient.scala [hunk 3]
```diff
@@ -307,7 +309,7 @@ private[spark] class AppClient(
    */
   def requestTotalExecutors(requestedTotal: Int): Boolean = {
     if (endpoint != null && appId != null) {
```
Comment (zsxwing) [inline core/src/main/scala/org/apache/spark/deploy/client/AppClient.scala]: endpoint.get != null && appId.get != null

File: core/src/main/scala/org/apache/spark/deploy/client/AppClient.scala [hunk 4]
```diff
@@ -283,19 +285,19 @@ private[spark] class AppClient(
 
   def start() {
     // Just launch an rpcEndpoint; it will call back into the listener.
-    endpoint = rpcEnv.setupEndpoint("AppClient", new ClientEndpoint(rpcEnv))
+    endpoint.set(rpcEnv.setupEndpoint("AppClient", new ClientEndpoint(rpcEnv)))
   }
 
   def stop() {
     if (endpoint != null) {
```
Comment (zsxwing) [inline core/src/main/scala/org/apache/spark/deploy/client/AppClient.scala]: endpoint.get != null

**Comment by zsxwing** [comment]
LGTM

**Comment by rxin** [comment]
I've merged this.

**Pull request closed without merging**